models/cice/model_mod.f90

```diff
@@ -1643,145 +1644,60 @@ end subroutine line_intercept
-   r(3) * x_corners(4)*y_corners(4)
-
-
-!----------------- Implementation test block
-! When interpolating on dipole x3 never exceeded 1e-9 error in this test
-!!!do i = 1, 4
-   !!!interp_val = a + r(1)*x_corners(i) + r(2)*y_corners(i)+ r(3)*x_corners(i)*y_corners(i)
-   !!!if(abs(interp_val - p(i)) > 1e-9) then
-      !!!write(*, *) 'large interp residual ', interp_val - p(i)
-   !!!endif
-!!!enddo
-!----------------- Implementation test block
-
-
-! Now do the interpolation
-expected_obs = a + r(1)*lon + r(2)*lat + r(3)*lon*lat
+! Performs IDW interpolation using great-circle distances for a quadrilateral.
 
-!********
-! Avoid exceeding maxima or minima as stopgap for poles problem
-! When doing bilinear interpolation in quadrangle, can get interpolated
-! values that are outside the range of the corner values
-if(expected_obs > maxval(p)) then
-   expected_obs = maxval(p)
-else if(expected_obs < minval(p)) then
-   expected_obs = minval(p)
-endif
-!********
-
-end subroutine quad_bilinear_interp
-
-!------------------------------------------------------------
+real(r8),  intent(in) :: lon, lat ! Interpolation point (longitude, latitude) in degrees
+real(r8),  intent(in) :: x_corners(4), y_corners(4) ! quadrilaterals corner points (longitude, latitude) in degrees.
+real(r8),  intent(in) :: p(4) ! values at the quadrilaterals corner points
+real(r8), intent(out) :: expected_obs ! Interpolated value at (lon, lat).
 
-subroutine mat3x3(m, v, r)
- real(r8),  intent(in) :: m(3, 3), v(3)
- real(r8), intent(out) :: r(3)
+! Set the power for the inverse distances
+real(r8), parameter :: power = 2.0_r8 ! Power for IDW (squared distance)
 
-! Solves rank 3 linear system mr = v for r
-! using Cramer's rule. This isn't the best choice
-! for speed or numerical stability so might want to replace
-! this at some point.
+! This value of epsilon radians is a distance of approximately 1 mm
+real(r8), parameter :: epsilon_radians = 1.56e-11_r8
 
-real(r8) :: m_sub(3, 3), numer, denom
-integer  :: i
+type(location_type) :: corner(4), point
+real(r8)            :: distances(4), inv_power_dist(4)
 
-! Compute the denominator, det(m)
-denom = deter3(m)
+integer :: i
 
-! Loop to compute the numerator for each component of r
-do i = 1, 3
-   m_sub = m
-   m_sub(:, i) = v   
-   numer = deter3(m_sub)
-   r(i) = numer / denom
+! Compute the distances from the point to each corner
+point = set_location(lon, lat, MISSING_R8, VERTISUNDEF)
+do i = 1, 4
+   corner(i) = set_location(x_corners(i), y_corners(i), MISSING_R8, VERTISUNDEF)
+   distances(i) = get_dist(point, corner(i), no_vert=.true.)
 enddo
 
-end subroutine mat3x3
-
-!------------------------------------------------------------
+if(minval(distances) < epsilon_radians) then
+   ! To avoid any round off issues, if smallest distance is less than epsilon radians
+   ! just assign the value at the closest gridpoint to the interpolant
+   expected_obs = p(minloc(distances,1))
```

Comment:
@jlaucar not sure if you want a return here, or you want to have this case also go through the test-out-of-bounds

Reply:
Let's go ahead and let it go through the out of bounds check. Obviously, it should never fail that, but this case is a boundary case that might not be that rare in some applications.

Reply:
I have completed a suite of tests on the ensemble version in my branch on this algorithm. I never violated the bounds and always appropriately fit the quad corners. I think this should now be tested in a real CICE case.

Reply:
Awesome, I ran this lastest code with the original out-of-bounds CICE reproduce  from Nuo's , and it completed successfully (no out-of-bounds error)